findimports.py

```diff
@@ -958,10 +961,12 @@ def printUnusedImports(self):
                         continue
                 print(f"{module.filename}:{lineno}: {name} not used")
 
-    def constructDot(self):
+    def constructDot(self, attributes=()):
         """Produce a dependency graph in dot format."""
         lines = list()
         lines.append("digraph ModuleDependencies {")
+        for a in attributes:
+            lines.append("  {};".format(a.rstrip(';')))
```

Comment:
So, I really want to handle the syntax at this point (add quotes around values, escape things that need to be escaped, etc).

Are graphviz attribute values typed?  E.g. is there a difference between attr=42 and attr="42"?  *looks it up*  No, there's no difference.

So we could always escape the value.  I already have a `quote()` function that I use to escape module.label.

Another possible source of error is if you use a keyword on the left of `=`.  Graphviz allows attributes with names that clash with keywords if you put them in quotes.  Do I need a smarter quote() function that leaves strings of the form `[a-zA-Z\200-\377_][a-zA-Z\200-\377_0-9]*` unquoted?

Also, what should happen if the user specifies -A something with no `=` sign, or -A x=y=z?  The latter is probably equivalent to `x="y=z"`.  I can easily imagine e.g. diagram titles that are like -A label="organization = total mess" or something.  (Also, there's no "title" attribute that I liked to use in previous examples for graphs, I [looked it up](https://graphviz.org/docs/graph/) -- but there's label and tooltip.)

Reply:
`-A something` with no `=` should maybe be treated as a boolean true?  E.g. `-A concentrate`, see https://graphviz.org/docs/attrs/concentrate/.

Or should we limit that to known boolean attributes and emit an error otherwise?  "Value missing for non-boolean attribute foo".

Reply:
> `-A something` with no `=` should maybe be treated as a boolean true? E.g. `-A concentrate`, see https://graphviz.org/docs/attrs/concentrate/.
> 
> Or should we limit that to known boolean attributes and emit an error otherwise? "Value missing for non-boolean attribute foo".

I think we should not allow automatic `true`. This will keep user more aware of `graphviz` syntax and keep things more in-line.

Reply:
> So, I really want to handle the syntax at this point (add quotes around values, escape things that need to be escaped, etc).

When input is generated by the code it makes sense.

Here I see one issue, what if user inputs:
```
-A 'label="word"'
```
How does one know if quotes are part of text or actual quotes?

Similar situation exists with
```
-A 'label=word;'
```
Do we treat it as part of "word;" or is it a separator?

Also, leaving literal placement will cause less confusion to the user.

In my experience, if I put too much processing for translation, eventually I am unable to figure out what input I need to get desired output.

I now tend to implement literal substitutions. It results in lower maintenance and being simpler.

I am leaning towards literal placement. If it is left how it is, then user can directly follow guidelines of `graphviz`. E.g.:
```
-A 'label="\"A";'
```
Works exactly as it would in:
https://dreampuf.github.io/GraphvizOnline/#digraph%20G%20%7B%0A%20%20label%20%3D%20%22%5C%22A%22%3B%0A%7D

Reply:
In short, I favour literal insertion of whatever user has provided. Together with explicit mention that this is what is being done.

If there was a robust and simple way of post-processing I wouldn't mind, but the path you are suggesting might lead into a rabbit hole and "consenting adults" approach might be better here.

Reply:
Your arguments mostly convince me.